jsettlers.logic/src/main/java/jsettlers/logic/map/grid/partition/manager/PartitionManager.java

```diff
@@ -380,9 +351,37 @@ private boolean tryToCreateWorker(WorkerCreationRequest workerCreationRequest) {
 
 		IManageableBearer manageableBearer = joblessBearer.removeObjectNextTo(workerCreationRequest.getPosition());
 		if(manageableBearer != null) {
-			return manageableBearer.becomeWorker(this, workerCreationRequest, offer);
-		} else {
-			return false;
+			if(manageableBearer.becomeWorker(this, workerCreationRequest, offer)) {
+				professionSettings.decrementBearerCount();
+				if(movableType == EMovableType.DIGGER) professionSettings.increamentDiggerCount();
+				if(movableType == EMovableType.BRICKLAYER) professionSettings.incrementBricklayerCount();
+				return true;
+			}
+		} 
+		return false;
+	}
+		
+	private void updateProfessionSettings() {
+		ProfessionSettings professionSettings = settings.getProfessionSettings();
+		professionSettings.resetCount();
+	
+		for (ILogicMovable mov : MovableManager.getAllMovables()) {
```

Comment:
I know that you just reused this code from my initial hack but wouldn't it make more sense to do this loop one time for all partitions instead of one iteration per partition every second?

Reply:
The more i am thinking about this i see the same problem here like with the Ui updaters ;) Every partition is managing its own timer instead of one timer that updates all partitions.

I agree with you that the MovableManager.getAllMovables() here feels a little bit wrong. But i am not sure how to separate this loop in a good way. The rest of the logic of updateProfessionSettings() should be moved to the ProfessionSettings class because it mainly operates on it. If the loop would be outside of PartitionManager then we would need methods that consume single moveables (like count(Moveable)). Another way would be to group the moveables by the partition and then pass all moveables of one partition at once. But i don't know if this is a good idea in the terms of performance.

Reply:
The main issue is that nobody really knows if and how long a movable stays on a partition. They could be converted, killed, pushed away (the biggest problem) or created at any time. We could hook different functions to see if their partition changes but this seams to be a too deep integration between partition and movable logic to me.


I thought about a new timerable in the PartitionsGrid that basically does:
for(mov in all movables) {

if(!mov.getMovableType().isPlayerControllable()) continue;

 partition = getPartitionAt(mov.getPosition);

if(mov.getPlayer().getPlayerId() != partition.playerId) continue;

settings = partition.getProfessionSettings();

switch(mov.getMovableType()) {
case BEARER:
settings.incBearerCount();
...
}

This needs just one iteration per movable and is not dependent on the number of partitions (in terms of complexity).

Reply:
> The main issue is that nobody really knows if and how long a movable stays on a partition. They could be converted, killed, pushed away (the biggest problem) or created at any time. We could hook different functions to see if their partition changes but this seams to be a too deep integration between partition and movable logic to me.
Totally agree. It would be very hard to not miss some case.

> I thought about a new timerable in the PartitionsGrid that basically does:
Yes and then remove the timer from each partition.

I will try to implement this.

Reply:
There is no real need to merge the execution of partitions:
- All tasks are executed in sequence (no synchronization between movable updater and partition update needed)
- its not any significant overhead for the timer
- its probably smaller code than iterating over every partitionObject
- it makes more sense for every partition to manage its own business

Reply:
You are right there are some things that could be a problem when looping centrally to update all partitions. So i will implement only the moveable loop in the PartitionGrid.

Reply:
It looks like this is not working. If i extract the logic to the grid then i face the following problems:
- Who is managing (starting/stopping) the timer ? Starting it in the constructor is not a good idea.
- It is necessary to iterate all partitions to reset them first (not a real problem but makes it a little bit less good)

So i will keep it like it is and just cleaned up the logic a little bit more. If you want you can also try but i think we can let it like it is for now.

Reply:
I was talking about an IScheduledTimerable as part of the global RescheduleTimer.
The RescheduleTimer gets saved and restored if you save/load a match and automatically started/setup or stopped when you open or close a match.

> It is necessary to iterate all partitions to reset them first (not a real problem but makes it a little bit less good)
I haven't thought about that but that shouldn't be an issue because there is the partitionObjects array in PartitionsGrid.

Reply:
Ah ok is see what my problem was. I was to much into the ui update logic. 

The RescheduleTimer.add(this, 1); should not be in the constructor. Adding something in constructor to global state is not a good idea. Instead the timer should be passed as paramter or the add should be externally.

In the loop the contains method of the PartitionManager could be reused.

Reply:
I agree that you shouldn't edit global state in the constructor but its not really the case here because the PartitionsGrid is only created in the constructor of MainGrid that one is only called to setup a new match. These constructors are just a better setup() function.

I think contains() would do mostly duplicate checks because we already know that the movable is not controllable and its part of the partition (and it would cost 3-4 more function calls per movable).